### PR TITLE
Replace IP with PG metadata db

### PR DESCRIPTION
**Pull request by depombo**
Introduces a singleton `MetadataRepository` that manages a long-lived `iasql_metadata` database bootstrapped when the `engine` starts. The entities and migrations live in `src/metadata`. `iasql` service takes the user ID directly instead of the entire `req.user` and `dbMan` provides default values for dbs when authentication is disabled.

### CODE REVIEW

File: src/metadata/entity/index.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+@Entity()
+export class IasqlUser {
+  // auth0 generated id
+  @PrimaryColumn()
+  id: string;
+
+  @Column()
+  email: string;
+
+  @ManyToMany(() => IasqlDatabase, {
```
Comment (depombo) [inline src/metadata/entity/index.ts:35]: made this many-to-many from the get-go for when we eventually want to have multiple users access the same database, but technically every database is associated with a single user atm

File: ormconfig-metadata.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
+const { SnakeNamingStrategy } = require('typeorm-naming-strategies');
```
Comment (aguillenv) [inline ormconfig-metadata.js:1]: Out of curiosity, why do we need this metadata config file? could not live in the current form config file?

**Review by aguillenv — COMMENTED** [review]
What I do not follow is the persistence of the tables after a restart or something? Or will they be part of the user database schemas? If so, they should be in a different schema and with some restrictions, right?

File: ormconfig-metadata.js (hunk 2 above)
Comment (dfellis) [reply ormconfig-metadata.js:1]: The "main" TypeORM migrations are for the tables we automatically put into every client database, while this is now metadata for the engine itself to manage the databases in its own private database, so we don't want those tables to show up in every user's databases.

File: ormconfig-metadata.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+module.exports = {
+ "type": "postgres",
+ "host": "localhost",
+ "port": 5432,
+ "username": "postgres",
+ "password": "test",
+ "database": "__example__",
```
Comment (dfellis) [inline ormconfig-metadata.js:9]: Do we want to use a different database name here?

File: src/router/db.ts [hunk 4]
```diff
@@ -14,13 +15,18 @@ db.post('/new', async (req, res) => {
     ].filter(k => !req.body.hasOwnProperty(k)).join(', ')}`
   );
   try {
-    res.json(await iasql.add(dbAlias, awsRegion, awsAccessKeyId, awsSecretAccessKey, req.user));
+    res.json(
+      await iasql.add(
+        dbAlias, awsRegion, awsAccessKeyId, awsSecretAccessKey, dbMan.getUid(req.user), dbMan.getEmail(req.user)
```
Comment (dfellis) [inline src/router/db.ts:20]: How are `dbMan.getUid` and `dbMan.getEmail` synchronous APIs?

File: src/services/db-manager.ts [hunk 5]
```diff
@@ -137,59 +123,14 @@ export function ourPgUrl(dbId: string): string {
...
-  const dbUser = metadata[toUserKey(dbAlias)];
-  delete metadata[toDbKey(dbAlias)];
-  delete metadata[toUserKey(dbAlias)];
-  await IronPlans.setTeamMetadata(ipUser.teamId, metadata);
-  return { dbId, dbUser, dbAlias };
-}
+// TODO type user
+export function getUid(user: any): string {
+  return config.a0Enabled ? user.sub : 'iasql';
+}
```
Comment (dfellis) [inline src/services/db-manager.ts:136]: Oh, these are literally the Auth0 values, that's how it can be synchronous.

File: src/services/repositories/metadata.ts [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+  private conn: Connection;
+  private userRepo: Repository<IasqlUser>;
+  private dbRepo: Repository<IasqlDatabase>;
+
+  async init() {
+    const conn = await createConnection(dbMan.baseConnConfig);
+    try {
+      await conn.query(`CREATE DATABASE ${this.database};`);
+    } catch(e) {
+      // postgres doesn't support IF NOT EXISTS for db creation so just ignore the error
```
Comment (dfellis) [inline src/services/repositories/metadata.ts:18]: You can avoid this try-catch with a conditional here. I think the following would work:

```plpgsql
DO $$
DECLARE
  db string;
BEGIN
  SELECT datname INTO db
  FROM pg_database
  WHERE datname = ${this.database};
  IF db IS NULL THEN
    CREATE DATABASE ${this.database};
  END IF;
END $$;
```

File: test.env [hunk 7]
```diff
@@ -0,0 +1,3 @@
+A0_ENABLED=true
+A0_DOMAIN=https://auth.iasql.com/
+A0_AUDIENCE=https://api.iasql.com
```
Comment (dfellis) [inline test.env:3]: Was this file committed by accident?

File: test/modules/aws-account-integration.ts [hunk 8]
```diff
@@ -14,8 +14,8 @@ const query = runQuery.bind(null, dbAlias);
 const sync = runSync.bind(null, dbAlias);
 
 jest.setTimeout(360000);
-beforeAll(execComposeUp);
-afterAll(() => execComposeDown());
+beforeAll(async () => await execComposeUp());
+afterAll(async () => await execComposeDown());
```
Comment (dfellis) [inline test/modules/aws-account-integration.ts:18]: This is a useless change. `async` makes the function return a promise, which is what `execComposeUp` already returns. You can revert all of these and it'll still work perfectly fine, just like it did before.

File: src/router/db.ts [hunk 9]
```diff
@@ -14,13 +15,18 @@ db.post('/new', async (req, res) => {
...
+      )
+    );
   } catch (e) {
     res.status(500).end(logger.error(e));
   }
 });
 
-db.post('/import', async (req, res) => {
+// TODO revive and test
```
[the remark this answers is not in the document]
Comment (depombo) [reply src/router/db.ts:28]: Commenting the import route for now? There is no way to test if changes to it are correct

File: test/modules/aws-account-integration.ts (hunk 8 above)
Comment (dfellis) [inline test/modules/aws-account-integration.ts:18]: In fact, since it doesn't use any of the input arguments either, wrapping them in anonymous functions is also not necessary. You can just write this and it should work fine:

```suggestion
beforeAll(execComposeUp);
afterAll(execComposeDown);
```

File: test/modules/aws-cloudwatch-integration.ts [hunk 10]
```diff
@@ -12,16 +12,16 @@ const uninstall = runUninstall.bind(null, dbAlias);
 
 const modules = ['aws_cloudwatch@0.0.1'];
 jest.setTimeout(240000);
-beforeAll(execComposeUp);
-afterAll(() => execComposeDown(modules));
+beforeAll(async () => await execComposeUp());
+afterAll(async () => await execComposeDown(modules));
```
Comment (dfellis) [inline test/modules/aws-cloudwatch-integration.ts:16]:
```suggestion
beforeAll(execComposeUp);
afterAll(execComposeDown.bind(null, modules));
```

File: test/modules/aws-account-integration.ts (hunk 8 above)
Comment (depombo) [reply test/modules/aws-account-integration.ts:18]: Well, no. `afterAll` was makes Jest hang up with an error that not all async operations completed when you run it for a single test. Feel free to try it

File: test/modules/aws-ecr-integration.ts [hunk 11]
```diff
@@ -15,16 +15,16 @@ const query = runQuery.bind(null, dbAlias);
 const modules = ['aws_ecr@0.0.1'];
 
 jest.setTimeout(240000);
-beforeAll(execComposeUp);
-afterAll(() => execComposeDown(modules));
+beforeAll(async () => await execComposeUp());
+afterAll(async () => await execComposeDown(modules));
```
Comment (dfellis) [inline test/modules/aws-ecr-integration.ts:19]:
```suggestion
beforeAll(execComposeUp);
afterAll(execComposeDown.bind(null, modules));
```

File: test/modules/aws-ecs-integration.ts [hunk 12]
```diff
@@ -45,16 +45,16 @@ const publicRepositoryName = `${prefix}${dbAlias}publicrepository`;
 const containerNamePublicRepository = `${prefix}${dbAlias}containerpublicrepository`;
 
 jest.setTimeout(240000);
-beforeAll(execComposeUp);
-afterAll(() => execComposeDown(modules));
+beforeAll(async () => await execComposeUp());
+afterAll(async () => await execComposeDown(modules));
```
Comment (dfellis) [inline test/modules/aws-ecs-integration.ts:49]:
```suggestion
beforeAll(execComposeUp);
afterAll(execComposeDown.bind(null, modules));
```

File: test/modules/aws-elb-integration.ts [hunk 13]
```diff
@@ -23,16 +23,16 @@ const lbType = LoadBalancerTypeEnum.APPLICATION;
 const lbIPAddressType = IpAddressType.IPV4;
 
 jest.setTimeout(360000);
-beforeAll(execComposeUp);
-afterAll(() => execComposeDown(modules));
+beforeAll(async () => await execComposeUp());
+afterAll(async () => await execComposeDown(modules));
```
Comment (dfellis) [inline test/modules/aws-elb-integration.ts:27]:
```suggestion
beforeAll(execComposeUp);
afterAll(execComposeDown.bind(null, modules));
```

File: src/services/repositories/metadata.ts (hunk 6 above)
Comment (depombo) [reply src/services/repositories/metadata.ts:18]: nice

File: test/modules/aws-vpc-integration.ts [hunk 14]
```diff
@@ -13,16 +13,16 @@ const availabilityZone = `${process.env.AWS_REGION ?? 'barf'}a`;
 const randIPBlock = Math.floor(Math.random() * 255);
 
 jest.setTimeout(240000);
-beforeAll(execComposeUp);
-afterAll(() => execComposeDown(modules));
+beforeAll(async () => await execComposeUp());
+afterAll(async () => await execComposeDown(modules));
```
Comment (dfellis) [inline test/modules/aws-vpc-integration.ts:17]:
```suggestion
beforeAll(execComposeUp);
afterAll(execComposeDown.bind(null, modules));
```

File: .github/workflows/ci.yml [hunk 15]
```diff
@@ -114,6 +114,5 @@ jobs:
           MODULES: ${{ needs.setup-module-test.outputs['test-modules'] }}
           AWS_ACCESS_KEY_ID: ${{ secrets[env.MATRIX_AWS_ACCESS_KEY_ID] }}
           AWS_SECRET_ACCESS_KEY: ${{ secrets[env.MATRIX_AWS_SECRET_ACCESS_KEY] }}
-          IRONPLANS_TOKEN: ${{ secrets.IRONPLANS_TOKEN }}
```
[the remark this answers is not in the document]
Comment (depombo) [reply .github/workflows/ci.yml:117]: that was an old token for auth0 authentication which is only needed if authentication is on. this test has authentication off. there was actually a bug in `simple-integration` (`AO` instead of `A0`) which that made it so that it never tested the authentication path so they IP tokens was completely ignored / misused

File: test/modules/aws-account-integration.ts (hunk 8 above)
Comment (dfellis) [reply test/modules/aws-account-integration.ts:18]: https://jestjs.io/docs/api#afterallfn-timeout

The docs do not mention an argument parameter, so it should work just fine. I'm guessing some other change happened at the same time and was misattributed to this change for fix?

Comment (depombo) [reply test/modules/aws-account-integration.ts:18]: https://github.com/iasql/iasql-engine/pull/542#discussion_r831152060

File: ormconfig-metadata.js (hunk 2 above)
Comment (aguillenv) [reply ormconfig-metadata.js:1]: oh so I think this answers part of my question [here](https://github.com/iasql/iasql-engine/pull/542#pullrequestreview-915462650) but how do we handle persistence? where are we creating and keeping our own private db with this info? I think that part is what I do not follow

Comment (dfellis) [reply ormconfig-metadata.js:1]: > oh so I think this answers part of my question [here](https://github.com/iasql/iasql-engine/pull/542#pullrequestreview-915462650) but how do we handle persistence? where are we creating and keeping our own private db with this info? I think that part is what I do not follow

That's why I asked about the accidental usage of `__example__` as the database name in another of my comments. :)

Our RDS database will have one database only accessible to the root account that we store this metadata in. (Similarly for anyone running it locally, it will be that same database name in their local postgres docker container.)

Basically we will have one singleton database that is for the engine itself and not for the users of the engine.

File: test/modules/aws-account-integration.ts (hunk 8 above)
Comment (depombo) [reply test/modules/aws-account-integration.ts:18]: this doesn't work

Comment (dfellis) [reply test/modules/aws-account-integration.ts:18]: Oh, that can explain that. They must be inspecting the type of the argument passed to it and doing different things for different types? :(

File: src/services/repositories/metadata.ts (hunk 6 above)
Comment (depombo) [reply src/services/repositories/metadata.ts:18]: actually I had tried a form of this and didn't because `QueryError: CREATE DATABASE cannot be executed from a function` going to leave a comment about that

**Review by dfellis — APPROVED** [review]

**Review by aguillenv — APPROVED** [review]
